### PR TITLE
feat: replace GraphQL Apollo with GraphQL Yoga

### PR DESCRIPTION
**Pull request by Moumouls**
### New Pull Request Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Issue Description

Related issue: #7936

### Approach

Remove apollo switch to Yoga with lazy loaded schema

### TODOs before merging

- [x] Add tests
- [x] Add changes to documentation (guides, repository pages, in-code descriptions)
- [x] Add [security check](https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md#security-checks)
- [x] Add new Parse Error codes to Parse JS SDK 
- [x] A changelog entry is created automatically using the pull request title (do not manually add a changelog entry)

### CODE REVIEW

File: src/GraphQL/loaders/filesMutations.js
Comment (ardatan) [inline src/GraphQL/loaders/filesMutations.js]: Have you tried to  use Buffer.from directly?

**Comment by Moumouls** [comment]
We need to waits for some fixes before merging:
- https://github.com/dotansimha/graphql-yoga/issues/1161
- https://github.com/dotansimha/graphql-yoga/issues/1162
- https://github.com/dotansimha/graphql-yoga/issues/1163

**Comment by Moumouls** [comment]
I also need to add an option to disable introspection or only authorize introspection with master key

**Comment by Moumouls** [comment]
After a quick K6 performance test, Yoga on Parse Server is able to handle GQL requests 30% faster.

Current alpha branch results:

```
     data_received..................: 9.3 MB 309 kB/s
     data_sent......................: 12 MB  391 kB/s
     http_req_blocked...............: avg=2.13µs  min=0s     med=1µs    max=2.38ms  p(90)=2µs     p(95)=2µs    
     http_req_connecting............: avg=92ns    min=0s     med=0s     max=331µs   p(90)=0s      p(95)=0s     
     http_req_duration..............: avg=9.61ms  min=6.12ms med=9.13ms max=75.08ms p(90)=12.25ms p(95)=13.41ms
       { expected_response:true }...: avg=9.61ms  min=6.12ms med=9.13ms max=75.08ms p(90)=12.25ms p(95)=13.41ms
     http_req_failed................: 0.00%  ✓ 0           ✗ 31026
     http_req_receiving.............: avg=21.55µs min=14µs   med=19µs   max=505µs   p(90)=30µs    p(95)=38µs   
     http_req_sending...............: avg=10.88µs min=5µs    med=10µs   max=801µs   p(90)=12µs    p(95)=14µs   
     http_req_tls_handshaking.......: avg=0s      min=0s     med=0s     max=0s      p(90)=0s      p(95)=0s     
     http_req_waiting...............: avg=9.58ms  min=6.09ms med=9.09ms max=75.02ms p(90)=12.22ms p(95)=13.38ms
     http_reqs......................: 31026  1033.932315/s
     iteration_duration.............: avg=9.66ms  min=6.16ms med=9.18ms max=77.58ms p(90)=12.3ms  p(95)=13.47ms
     iterations.....................: 31026  1033.932315/s
     vus............................: 10     min=10        max=10 
     vus_max........................: 10     min=10        max=10 
```

This branch with Yoga

```
     data_received..................: 11 MB 358 kB/s
     data_sent......................: 15 MB 514 kB/s
     http_req_blocked...............: avg=2.26µs  min=0s     med=1µs    max=4ms      p(90)=2µs    p(95)=2µs   
     http_req_connecting............: avg=81ns    min=0s     med=0s     max=414µs    p(90)=0s     p(95)=0s    
     http_req_duration..............: avg=7.29ms  min=4.02ms med=6.99ms max=98.15ms  p(90)=9.04ms p(95)=9.83ms
       { expected_response:true }...: avg=7.29ms  min=4.02ms med=6.99ms max=98.15ms  p(90)=9.04ms p(95)=9.83ms
     http_req_failed................: 0.00% ✓ 0           ✗ 40829
     http_req_receiving.............: avg=20.16µs min=13µs   med=18µs   max=484µs    p(90)=25µs   p(95)=35µs  
     http_req_sending...............: avg=10.5µs  min=5µs    med=10µs   max=838µs    p(90)=12µs   p(95)=13µs  
     http_req_tls_handshaking.......: avg=0s      min=0s     med=0s     max=0s       p(90)=0s     p(95)=0s    
     http_req_waiting...............: avg=7.26ms  min=3.95ms med=6.95ms max=98.11ms  p(90)=9.01ms p(95)=9.8ms 
     http_reqs......................: 40829 1360.673351/s
     iteration_duration.............: avg=7.34ms  min=4.15ms med=7.03ms max=102.31ms p(90)=9.09ms p(95)=9.88ms
     iterations.....................: 40829 1360.673351/s
     vus............................: 10    min=10        max=10 
     vus_max........................: 10    min=10        max=10 
```

The K6 test is a simple query on users. this query does not return results. The K6 script shows the pure performance of the GraphQL Parse Server layer.

File: src/GraphQL/loaders/filesMutations.js
Comment (Moumouls) [reply src/GraphQL/loaders/filesMutations.js]: Thanks @ardatan  for the suggestion, it works.

File: package.json
Comment (davimacedo) [inline package.json:101]: Why do we need to downgrade it?

File: src/GraphQL/ParseGraphQLServer.js
Comment (davimacedo) [inline src/GraphQL/ParseGraphQLServer.js:60]: I am afraid of initializing the entire server for every schema change. I took a quick look on yoga repo and I saw a lot of things being initialized in here. Isn't there any other way?

File: spec/ParseGraphQLServer.spec.js
Comment (davimacedo) [inline spec/ParseGraphQLServer.spec.js]: Is it a breaking change?

File: spec/ParseGraphQLServer.spec.js
Comment (davimacedo) [inline spec/ParseGraphQLServer.spec.js]: Is it a breaking change?

File: src/GraphQL/ParseGraphQLServer.js
Comment (davimacedo) [inline src/GraphQL/ParseGraphQLServer.js:101]: I'd probably address this on a separate pr

File: src/GraphQL/loaders/defaultGraphQLTypes.js
Comment (davimacedo) [inline src/GraphQL/loaders/defaultGraphQLTypes.js:225]: Do you think that the type change could be a breaking change for someone with using auto or maybe customizing the schema?

**Review by davimacedo — COMMENTED** [review]
Thanks for the PR! I've gone through the entire issue discussion and it sound promising. Please find some comments below.

File: package.json
Comment (mtrezza) [reply package.json:101]: I don't think we should, see https://github.com/parse-community/parse-server/issues/7978. The package is used only by a GraphQL test file and there seems to an incompatibility, I have disabled the 2 tests for now in https://github.com/parse-community/parse-server/pull/7896, it's an open issue to fix this or replace node-fetch.

Comment (Moumouls) [reply package.json:101]: > I have disabled the 2 tests for now

I think it's not a good idea, Multipart File Upload is an important feature of the GQL API. It could definitely break many parse-servers deployment, this feature MUST be validated by the CI @mtrezza

File: src/GraphQL/ParseGraphQLServer.js
Comment (Moumouls) [reply src/GraphQL/ParseGraphQLServer.js:60]: @davimacedo I added a test to ensure that the server is only recreated only when schema change is detected (should not occur often). Also, the K6 benchmark shows that performance is better.

I discussed with the maintainers of Yoga, that a new option will be added in the future to handle better this use. But for now, it works pretty well. 🙂

File: spec/ParseGraphQLServer.spec.js
Comment (Moumouls) [reply spec/ParseGraphQLServer.spec.js]: No this is a security enhancement, also discussed with The Guild maintainers.
The server still accepts `*` but the server auto-detect the incoming origin and  "scope" the response to this origin. It's an improved security mechanism included out of the box in Yoga.

File: spec/ParseGraphQLServer.spec.js
Comment (Moumouls) [reply spec/ParseGraphQLServer.spec.js]: No, a missing test, from the previous GQL API, the max file size was not tested on Upload. I added the test.

File: src/GraphQL/ParseGraphQLServer.js
Comment (Moumouls) [reply src/GraphQL/ParseGraphQLServer.js:101]: The playground is now not maintained anymore. This is why yoga exposes the `renderGraphiQL`.

We also already have graphiQL on the parse dashboard: https://github.com/parse-community/parse-dashboard/blob/e43ba0bf18aa2e951b6dd22fcaf73e961cd0c780/package.json#L47

Better DX since developers will have the same GQL IDE on both projects.

ref: https://github.com/graphql/graphql-playground/issues/1366#issuecomment-1062088978

File: src/GraphQL/loaders/defaultGraphQLTypes.js
Comment (Moumouls) [reply src/GraphQL/loaders/defaultGraphQLTypes.js:225]: Not a breaking change, this is a dummy type. Before we used the GraphQLUpload type from the `graphql-upload` package. This package is now removed since Yoga support multipart GQL upload out of the box.

File: package.json
Comment (mtrezza) [reply package.json:101]: If the feature must be validated then I suggest we fix the node-fetch issue. I would say security outweighs a potential bug. Remember, 2.x has been upgraded to 3.x. due to a security vulnerability. It's also a difficult case to make that every deployment who doesn't use GraphQL should accept a security vulnerability because of GraphQL. GraphQL needs to finally move out of Parse Server, and as long as it doesn't it cannot compromise the security of Parse Server as a whole.

File: spec/ParseGraphQLServer.spec.js
Comment (mtrezza) [reply spec/ParseGraphQLServer.spec.js]: Please use `example.com` to not violate RFC.

File: package.json
Comment (Moumouls) [reply package.json:101]: @davimacedo  `node-fetch` is only used in the ParseGraphQL.Server test file, it's safe to downgrade to 2.X until a PR could be sent with the 3.X support or a complete refactor with another lib.

Comment (Moumouls) [reply package.json:101]: > Remember, 2.x has been upgraded to 3.x. due to a security vulnerability.

@mtrezza I understand perfectly, but here the potential security issue only affects the parse server test localhost test suite. I do not agree to disable tests related to important features used by the community for an internal dev package version issue with low-security risk, since this package is only a dev dep.

I'm may be wrong but the breaking change from 2.X to 3.X on node fetch is related to the switch to ESM. I discussed this also with The Guild.

I see that the upgrade was created  here https://github.com/parse-community/parse-server/pull/7782

And also a security update was also released on the 2.X: https://github.com/node-fetch/node-fetch/releases/tag/v2.6.7

Capture d’écran 2022-05-06 à 20 41 28

The 2.X has also a LTS it seems. @mtrezza

Comment (mtrezza) [reply package.json:101]: > since this package is only a dev dep.

We have seen major security vulnerabilities in the past due to dev dependencies. Security goes beyond a deployed server, it includes the CI environment as well. Think of private repositories, proprietary code, environment secrets, etc. The package is there and we cannot know what else it will be used for in other tests or more generally in a CI/CD pipeline.

> The 2.X has also a LTS it seems

Until when? If they offer LTS for 2.x then I'd be OK with going back to 2.x for now.

Comment (Moumouls) [reply package.json:101]: form npm page

```
node-fetch from v3 is an ESM-only module - you are not able to import it with require().

If you cannot switch to ESM, please use v2 which remains compatible with CommonJS. Critical bug fixes will continue to be published for v2.
```

Not mentioned, an issue could be opened on the node-fetch repo.

Anyway this is just my opinion, I don't think it's a reasonable solution to disable the tests compared to the current situation. in any case I will keep node-fetch 2.X on my fork to ensure that this feature works. I don't think I'll investigate further on my side a PR to switch back to node 3.X or another fetch package.

@mtrezza

Comment (mtrezza) [reply package.json:101]: Yes, I think we could just ask them for a LTS end date and downgrade to 2.x. After that date we'll have to upgrade anyway, because we are just moving the debt - a strategy that we pursued in Parse Dashboard for a long time and has eventually caused a lot of headache.

I've opened https://github.com/node-fetch/node-fetch/discussions/1553.

A compromise may be to stay on 3.x but enable these 2 tests only for <Node 17, so they run at least for lower node versions for now. That may help to keep an eye on the feature. Node 16 LTS ends Apr 2024, so we'd run the 2 tests for a long time. Meanwhile we keep https://github.com/parse-community/parse-server/issues/7978 open to not forget about it.

Let's wait what node-fetch answers, but the compromise seems more attractive to me I think.

Comment (mtrezza) [reply package.json:101]: See https://github.com/node-fetch/node-fetch/discussions/1553#discussioncomment-2703103.

They already released beta versions of 4.x, plus people are likely switching to Node's built-in `fetch` which is available from Node 18. I think we cannot be sure how long and how well maintained 2.x will stay. Their maintenance model seems to be more community driven than policy driven, so it's hard to tell.

Maybe let's try to go with the compromise:
- Stay on 3.x and run the tests only for Node <17
- Do not run the tests on Node 17, which is an odd-intermediate version where LTS ends 2022-06 anyway
- Over time (until Apr 2024), create a PR to run the tests on Node 18 using Node's built-in `fetch`

I have re-enabled the tests for Node <17 in https://github.com/parse-community/parse-server/pull/7980 and we keep track to enable the tests for Node 18 in https://github.com/parse-community/parse-server/issues/7978.

**Comment by Moumouls** [comment]
ready to merge @mtrezza  🚀

**Comment by mtrezza** [comment]
Could you please coordinate with the reviewers to resolve the open comments?

File: package.json
Comment (ardatan) [reply package.json:101]: You can use `cross-undici-fetch`, that uses the best available version of `fetch` (global/native one if available, node-fetch or undici). It also has implementations for `FormData` ,  `File` ,  `ReadableStream` etc...

Comment (Moumouls) [reply package.json:101]: @ardatan I tried the switch to cross-undici-fetch but new fails appeared. The switch from node-fetch is out of the scope of this PR from my point of view.

Feel free to open a new PR to switch replace node-fetch by cross-undici-fetch you will maybe able to fix fails more easily and quickly than me. 🙂

**Comment by Moumouls** [comment]
@davimacedo  if you have some time to take a look to this one ?

**Review by davimacedo — APPROVED** [review]
LGTM!

**Review by mtrezza — COMMENTED** [review]
Last question before we merge this, is there any migration needed on the developer side for this PR?

**Comment by Moumouls** [comment]
@mtrezza no migration needed, GraphQL API is standard and Yoga follow GraphQL specs like Apollo ! 

We are good to go, i also tested this feature on my company projects.

**Comment by mtrezza** [comment]
Amazing, let's merge then.

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.3.0-alpha.11](https://github.com/parse-community/parse-server/releases/tag/5.3.0-alpha.11)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.3.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/5.3.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.3.0](https://github.com/parse-community/parse-server/releases/tag/5.3.0)

**Pull request merged**
